src/lib/ppx_coda/versioned.ml

```diff
@@ -0,0 +1,114 @@
+(* versioned.ml -- static enforcement of versioned types via ppx
+
+   1) check that versioned type always in Stable.Vn.T module hierarchy
+   2) versioned types depend only on other versioned types (or OCaml built-in types)
+
+  to use, add coda_ppx to the dune pps list, and add "versioned" to deriving list for a type
+
+ *)
+
+open Core_kernel
+open Ppxlib
+
+let is_versioned_ident item =
+  match item with
+  | Pexp_ident id -> (
+    match id.txt with Lident s -> String.equal s "versioned" | _ -> false )
+  | _ -> false
+
+let payload_has_versioned payload =
+  match payload with
+  | PStr structure ->
+      List.exists structure ~f:(fun str ->
+          match str.pstr_desc with
+          | Pstr_eval (expr, _) -> (
+            (* the "versioned" can appear as a singleton ident, or in a tuple *)
+            match expr.pexp_desc with
+            | Pexp_ident _ -> is_versioned_ident expr.pexp_desc
+            | Pexp_tuple items ->
+                List.exists items ~f:(fun item ->
+                    is_versioned_ident item.pexp_desc )
+            | _ -> false )
+          | _ -> false )
+  | _ -> false
+
+let attribute_has_deriving_versioned ((name, payload) : attribute) =
+  String.equal name.txt "deriving" && payload_has_versioned payload
+
+let attributes_have_deriving_versioned (attrs : attribute list) =
+  List.exists attrs ~f:attribute_has_deriving_versioned
+
+let type_has_deriving_versioned type_decl =
+  attributes_have_deriving_versioned type_decl.ptype_attributes
+
+(* check that a versioned type occurs in valid module hierarchy and is named "t" *)
+let validate_versioned type_decl inner3_modules =
+  if type_has_deriving_versioned type_decl then (
+    if not (String.equal type_decl.ptype_name.txt "t") then
+      Location.raise_errorf ~loc:type_decl.ptype_loc
+        "Versioned type must be named \"t\", got: \"%s\""
+        type_decl.ptype_name.txt ;
+    match inner3_modules with
+    | ["T"; version_module; "Stable"] ->
+        let len = String.length version_module in
+        if (not (Char.equal version_module.[0] 'V')) && len > 1 then
```

Comment:
This looks like it lets through any module with a single digit name. I think the `&& len > 1` might want to be a `|| len <= 1` instead.

Reply:
Yup, wrong parenthesization.